contrib/internal/redigo/redigo.go

```diff
@@ -0,0 +1,92 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+
+package redigo
+
+import (
+	"net/url"
+	"reflect"
+	"regexp"
+	"strconv"
+	"unsafe"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+
+	garyburdredis "github.com/garyburd/redigo/redis"
+	gomoduleredis "github.com/gomodule/redigo/redis"
+)
+
+// RedisDialConfig contains the configurable parameters for the Redis connection.
+type RedisDialConfig struct {
+	DB int
+}
+
+const (
+	structFieldNameDialOptionFunc = "f"
+	structFieldNameConfigDBIndex  = "db"
+)
+
+// DialOption is a type constraint for the redis.DialOption struct from both redigo integrations.
+type DialOption interface {
+	garyburdredis.DialOption | gomoduleredis.DialOption
+}
+
+// ResolveDialOpts resolves the config from the given []redis.DialOption.
+// Since both the config struct (redis.dialOption) and the redis.DialOption.f struct field are unexported,
+// this function uses reflection to initialize and access the struct fields.
+func ResolveDialOpts[T DialOption](opts []T) (*RedisDialConfig, bool) {
+	if len(opts) == 0 {
+		return &RedisDialConfig{}, true
+	}
+	defer func() {
+		if err := recover(); err != nil {
+			log.Debug("contrib/internal/redigo: Failed to resolve dial options")
+		}
+	}()
+	cfg := &RedisDialConfig{}
```

Comment:
If you're going to declare this, then just do it in the first line of the function. Then you can just return that config if there are no options. e.g.

```
	cfg := &RedisDialConfig{}
	if len(opts) == 0 {
		return cfg, true
	}
```